Convert driver exceptions when starting transactions

lcobucci opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | bug

#### Summary

The wrapped connection may no longer be valid for long-running processes, and starting a transaction will cause driver exceptions.

DBAL Connection wasn't converting these, leaking internal details of the platform and not providing a consistent API (all other methods would throw a `ConnectionLost` exception for the scenario above).

This ensures we do have the expected behaviour for this edge case.

lcobucci committed:
Convert driver exceptions when starting transactions

For long-running processes, it's possible that the wrapped connection is
no longer valid and starting a transaction will lead to driver
exceptions.

DBAL Connection wasn't converting these, leaking internal details of the
platform and not providing a consistent API (all other methods would
throw a `ConnectionLost` exception for the scenario above).

This ensures we do have the expected behaviour for this edge-case.

Signed-off-by: Luís Cobucci <lcobucci@gmail.com>

morozov commented:
Let's get it integration-tested. Even though the unit test passes, and the code works as expected on MySQL, we want to be sure that it works the same on other platforms as well or at least know that it doesn't.

lcobucci committed:
Kill connection instead of waiting for idle timeout

This optimises how this test is executed by leveraging two connections
and killing the one executing the tests.

It also aims at making the whole conversion process available to other
platforms.

Signed-off-by: Luís Cobucci <lcobucci@gmail.com>

lcobucci committed:
Ensure connection lost detection is available for PGSQL

This extends the functional test for connection loss detection to verify
the behaviour against Postgres and ensures the ExceptionConverter for
that API behaves as expected.

Signed-off-by: Luís Cobucci <lcobucci@gmail.com>

lcobucci commented on tests/Functional/TransactionTest.php:
```diff
@@ -65,6 +62,24 @@ private function expectConnectionLoss(callable $scenario): void
+            default => self::markTestSkipped('Unsupported test platform.'),
+        };
+
+        $privilegedConnection = TestUtil::getPrivilegedConnection();
```
Using priviledged connection here, assuming that's a requirement for other platforms (maybe oracle or SQL server?)

lcobucci commented on tests/Functional/TransactionTest.php:
```diff
@@ -9,43 +9,40 @@
 {
     public function testCommitFailure(): void
     {
+        $this->markConnectionNotReusable();
```
Marking as not reusable as I didn't want for a failure in one test to leak into the other

lcobucci commented on src/Driver/API/PostgreSQL/ExceptionConverter.php:
```diff
@@ -77,6 +78,13 @@ public function convert(Exception $exception, ?Query $query): DriverException
                 return new ConnectionException($exception, $query);
         }
 
+        if (
+            $exception->getSQLState() === 'HY000'
+            && str_contains($exception->getMessage(), 'server closed the connection')
+        ) {
+            return new ConnectionLost($exception, $query);
+        }
```
This might only be applicable for recent PGSQL versions...

lcobucci replied:
Nah... only PDO has the SQL State set for this error :)

Ooohh I missed working with DBAL :sweat_smile:

lcobucci commented on tests/Functional/TransactionTest.php:
```diff
@@ -21,10 +21,21 @@
 
 class TransactionTest extends FunctionalTestCase
 {
+    public function testBeginTransactionFailure(): void
```
And, obviously, `mysqli` behaves differently....

lcobucci replied:
75ced17bc addresses that

lcobucci committed:
Propagate mysqli errors when starting transactions

This is required because the MySQLi API only returns a boolean result
that we were simply not using.

Signed-off-by: Luís Cobucci <lcobucci@gmail.com>

morozov commented on src/Connection.php:
```diff
@@ -1026,10 +1026,16 @@ public function beginTransaction(): void
 
         ++$this->transactionNestingLevel;
 
-        if ($this->transactionNestingLevel === 1) {
```
I wouldn't rearrange this code. Non-nested transactions are more common, and you just wanted to wrap the call to `$connection->beginTransaction()` in a try/catch. Let's not change anything else.

lcobucci replied:
Code structure reverted

morozov commented on tests/Functional/TransactionTest.php:
```diff
@@ -7,45 +7,55 @@
 {
+    public function testBeginTransactionFailure(): void
+    {
+        $this->markConnectionNotReusable();
```
Would it be fair to also move this call to `killCurrentSession()`? This way, there will be less code duplication and be even more obvious why it's not reusable.

morozov commented on tests/Functional/TransactionTest.php:
```diff
@@ -65,6 +75,31 @@ private function expectConnectionLoss(callable $scenario): void
+
+        $privilegedConnection = TestUtil::getPrivilegedConnection();
+        $privilegedConnection->executeStatement(
+            sprintf(
```
Please try passing the PID as a statement parameter.

morozov reviewed (commented):
The code change and the tests look good. Thanks for replacing `sleep()` with killing the session. Just a few nits inline.

morozov commented:
Thanks, @lcobucci!

lcobucci commented:
@morozov @derrabus would it be alright to release 4.2.3 soon or are you planning to group more items under it?

morozov commented:
I'm fine with releasing it any time.

lcobucci commented:
> I'm fine with releasing it any time.

It would be great to have it released, then
